File: ruby-sdk/lib/cleanster/api/properties_api.rb
module Cleanster
  module Api
    # PropertiesApi - CRUD for cleaning locations, plus cleaners, iCal, and checklists.
    class PropertiesApi
      def initialize(http)
        @http = http
      end

      # List all properties, optionally filtered by service type.
      #
      # @param service_id [Integer, nil]
      # @return [Models::ApiResponse]
      def list_properties(service_id: nil)
        params = service_id ? { serviceId: service_id } : nil
        raw    = @http.get("/v1/properties", params: params)
        Models::ApiResponse.from_hash(raw)
      end

      # Add a new property.
      #
      # @param request [Hash] name, optional nick_name, address, city, country,
      #   room_count, bathroom_count, service_id.
      # @return [Models::ApiResponse<Models::Property>]
      def add_property(request)
        raw = @http.post("/v1/properties", body: camel_keys(request))
        Models::ApiResponse.from_hash(raw, model_class: Models::Property)
      end

      # Get details of a specific property.
      #
      # @param property_id [Integer]
      # @return [Models::ApiResponse<Models::Property>]
      def get_property(property_id)
        raw = @http.get("/v1/properties/#{property_id}")
        Models::ApiResponse.from_hash(raw, model_class: Models::Property)
      end

      # Update an existing property.
      #
      # @param property_id [Integer]
      # @param request     [Hash]
      # @return [Models::ApiResponse<Models::Property>]
      def update_property(property_id, request)
        raw = @http.put("/v1/properties/#{property_id}", body: camel_keys(request))
        Models::ApiResponse.from_hash(raw, model_class: Models::Property)
      end

      # Update additional information fields for a property.
      def update_additional_information(property_id, data)
        raw = @http.put("/v1/properties/#{property_id}/additional-information", body: data)
        Models::ApiResponse.from_hash(raw)
      end

      # Enable or disable a property.
      #
      # @param property_id [Integer]
      # @param enabled     [Boolean]
      # @return [Models::ApiResponse]
      def enable_or_disable_property(property_id, enabled:)
        raw = @http.post("/v1/properties/#{property_id}/enable-disable", body: { enabled: enabled })
        Models::ApiResponse.from_hash(raw)
      end

      # Permanently delete a property.
      def delete_property(property_id)
        raw = @http.delete("/v1/properties/#{property_id}")
        Models::ApiResponse.from_hash(raw)
      end

      # Get the list of cleaners assigned to a property.
      def get_property_cleaners(property_id)
        raw = @http.get("/v1/properties/#{property_id}/cleaners")
        Models::ApiResponse.from_hash(raw)
      end

      # Assign a cleaner to a property's default cleaner pool.
      #
      # @param property_id [Integer]
      # @param cleaner_id  [Integer]
      # @return [Models::ApiResponse]
      def assign_cleaner_to_property(property_id, cleaner_id:)
        raw = @http.post("/v1/properties/#{property_id}/cleaners", body: { cleanerId: cleaner_id })
        Models::ApiResponse.from_hash(raw)
      end

      # Remove a cleaner from a property's default cleaner pool.
      def unassign_cleaner_from_property(property_id, cleaner_id)
        raw = @http.delete("/v1/properties/#{property_id}/cleaners/#{cleaner_id}")
        Models::ApiResponse.from_hash(raw)
      end

      # Add one or more iCal calendar links to a property for availability syncing.
      # Each URL must be a live, publicly fetchable .ics feed - the API validates
      # the feed content, not just the URL shape.
      #
      # @param property_id    [Integer]
      # @param calendar_links [Array<String>, String] One or more iCal feed URLs.
      # @return [Models::ApiResponse]
      def add_ical_link(property_id, calendar_links:)
        links = Array(calendar_links)
        raw = @http.put("/v1/properties/#{property_id}/ical", body: { calendarLinks: links })
        Models::ApiResponse.from_hash(raw)
      end

      # Get the calendar links currently attached to a property.
      # Returns data as an array of { "id" => Integer, "calendarLink" => String }.
      def get_ical_link(property_id)
        raw = @http.get("/v1/properties/#{property_id}/ical")
        Models::ApiResponse.from_hash(raw)
      end

      # Remove one or more calendar links from a property, by numeric link ID
      # (from get_ical_link) - not by URL.
      #
      # @param property_id [Integer]
      # @param ids         [Array<Integer>, Integer] The link ID(s) to remove.
      def remove_ical_link(property_id, ids:)
        link_ids = Array(ids)
        raw = @http.delete("/v1/properties/#{property_id}/ical", body: { ids: link_ids })
        Models::ApiResponse.from_hash(raw)
      end

      # Set a default checklist for a property.
      #
      # @param property_id               [Integer]
      # @param checklist_id              [Integer]
      # @param update_upcoming_bookings  [Boolean]
      # @return [Models::ApiResponse]
      def set_default_checklist(property_id, checklist_id, update_upcoming_bookings: false)
        path = "/v1/properties/#{property_id}/checklist/#{checklist_id}" \
               "?updateUpcomingBookings=#{update_upcoming_bookings}"
        raw = @http.put(path)
        Models::ApiResponse.from_hash(raw)
      end

      private

      def camel_keys(hash)
        map = {
          room_count:     "roomCount",
          bathroom_count: "bathroomCount",
          service_id:     "serviceId"
        }
        hash.transform_keys { |k| map[k] || k }
      end
    end
  end
end

File: php-sdk/src/Api/PropertiesApi.php
<?php

declare(strict_types=1);

namespace Cleanster\Api;

use Cleanster\HttpClient;
use Cleanster\Models\ApiResponse;
use Cleanster\Models\Property;

/**
 * Manages physical cleaning locations.
 */
final class PropertiesApi
{
    public function __construct(private readonly HttpClient $http) {}

    /**
     * Return all properties, optionally filtered by service type.
     *
     * @param int|null $serviceId Pass null to return all service types.
     * @return ApiResponse<Property[]>
     */
    public function listProperties(?int $serviceId = null): ApiResponse
    {
        $query = [];
        if ($serviceId !== null) {
            $query['serviceId'] = $serviceId;
        }
        $raw   = $this->http->get('/v1/properties', $query);
        $items = array_map(fn(array $p) => new Property($p), $raw['data'] ?? []);
        return new ApiResponse($raw['status'] ?? 200, $raw['message'] ?? 'OK', $items);
    }

    /**
     * Add a new property to the partner account.
     *
     * @return ApiResponse<Property>
     */
    public function addProperty(array $request): ApiResponse
    {
        $raw = $this->http->post('/v1/properties', $request);
        return new ApiResponse($raw['status'] ?? 200, $raw['message'] ?? 'OK', new Property($raw['data'] ?? []));
    }

    /**
     * Return details of a specific property.
     *
     * @return ApiResponse<Property>
     */
    public function getProperty(int $propertyId): ApiResponse
    {
        $raw = $this->http->get("/v1/properties/{$propertyId}");
        return new ApiResponse($raw['status'] ?? 200, $raw['message'] ?? 'OK', new Property($raw['data'] ?? []));
    }

    /**
     * Replace all fields of an existing property.
     *
     * @return ApiResponse<Property>
     */
    public function updateProperty(int $propertyId, array $request): ApiResponse
    {
        $raw = $this->http->put("/v1/properties/{$propertyId}", $request);
        return new ApiResponse($raw['status'] ?? 200, $raw['message'] ?? 'OK', new Property($raw['data'] ?? []));
    }

    /** Update freeform additional information for a property. */
    public function updateAdditionalInformation(int $propertyId, array $data): ApiResponse
    {
        $raw = $this->http->put("/v1/properties/{$propertyId}/additional-information", $data);
        return new ApiResponse($raw['status'] ?? 200, $raw['message'] ?? 'OK', $raw['data'] ?? []);
    }

    /** Toggle a property's active state. */
    public function enableOrDisableProperty(int $propertyId, bool $enabled): ApiResponse
    {
        $raw = $this->http->post("/v1/properties/{$propertyId}/enable-disable", ['enabled' => $enabled]);
        return new ApiResponse($raw['status'] ?? 200, $raw['message'] ?? 'OK', $raw['data'] ?? []);
    }

    /** Permanently delete a property. */
    public function deleteProperty(int $propertyId): ApiResponse
    {
        $raw = $this->http->delete("/v1/properties/{$propertyId}");
        return new ApiResponse($raw['status'] ?? 200, $raw['message'] ?? 'OK', $raw['data'] ?? []);
    }

    /** Return all cleaners assigned to a property. */
    public function getPropertyCleaners(int $propertyId): ApiResponse
    {
        $raw = $this->http->get("/v1/properties/{$propertyId}/cleaners");
        return new ApiResponse($raw['status'] ?? 200, $raw['message'] ?? 'OK', $raw['data'] ?? []);
    }

    /** Add a cleaner to a property's default cleaner pool. */
    public function assignCleanerToProperty(int $propertyId, int $cleanerId): ApiResponse
    {
        $raw = $this->http->post("/v1/properties/{$propertyId}/cleaners", ['cleanerId' => $cleanerId]);
        return new ApiResponse($raw['status'] ?? 200, $raw['message'] ?? 'OK', $raw['data'] ?? []);
    }

    /** Remove a cleaner from a property's default cleaner pool. */
    public function unassignCleanerFromProperty(int $propertyId, int $cleanerId): ApiResponse
    {
        $raw = $this->http->delete("/v1/properties/{$propertyId}/cleaners/{$cleanerId}");
        return new ApiResponse($raw['status'] ?? 200, $raw['message'] ?? 'OK', $raw['data'] ?? []);
    }

    /**
     * Add one or more iCal calendar links to a property for availability syncing.
     * Each URL must be a live, publicly fetchable .ics feed - the API validates
     * the feed content, not just the URL shape.
     *
     * @param string[]|string $calendarLinks One or more iCal feed URLs.
     */
    public function addICalLink(int $propertyId, array|string $calendarLinks): ApiResponse
    {
        $links = is_array($calendarLinks) ? $calendarLinks : [$calendarLinks];
        $raw = $this->http->put("/v1/properties/{$propertyId}/ical", ['calendarLinks' => $links]);
        return new ApiResponse($raw['status'] ?? 200, $raw['message'] ?? 'OK', $raw['data'] ?? []);
    }

    /** Return the calendar links currently attached to a property, as a list of ['id' => int, 'calendarLink' => string]. */
    public function getICalLink(int $propertyId): ApiResponse
    {
        $raw = $this->http->get("/v1/properties/{$propertyId}/ical");
        return new ApiResponse($raw['status'] ?? 200, $raw['message'] ?? 'OK', $raw['data'] ?? []);
    }

    /**
     * Remove one or more calendar links from a property, by numeric link ID
     * (from getICalLink) - not by URL.
     *
     * @param int[]|int $ids The link ID(s) to remove.
     */
    public function removeICalLink(int $propertyId, array|int $ids): ApiResponse
    {
        $linkIds = is_array($ids) ? $ids : [$ids];
        $raw = $this->http->delete("/v1/properties/{$propertyId}/ical", ['ids' => $linkIds]);
        return new ApiResponse($raw['status'] ?? 200, $raw['message'] ?? 'OK', $raw['data'] ?? []);
    }

    /**
     * Set the default checklist for a property.
     *
     * @param bool $updateUpcomingBookings If true, also applies to all future bookings at this property.
     */
    public function setDefaultChecklist(
        int  $propertyId,
        int  $checklistId,
        bool $updateUpcomingBookings = false
    ): ApiResponse {
        $path = "/v1/properties/{$propertyId}/checklist/{$checklistId}"
              . '?updateUpcomingBookings=' . ($updateUpcomingBookings ? 'true' : 'false');
        $raw  = $this->http->put($path);
        return new ApiResponse($raw['status'] ?? 200, $raw['message'] ?? 'OK', $raw['data'] ?? []);
    }
}
